lib/srv/desktop/audit_cache.go

```diff
@@ -0,0 +1,209 @@
+
+const (
+	// entryMaxItems is the maximum number of items we want
+	// to allow in a single sharedDirectoryAuditCacheEntry.
+	//
+	// It's not a precise value, just one that should give us
+	// substantial wiggle room under normal operating conditions,
+	// but will thwart a hostile client attempting a memory DOS attack.
+	// https://github.com/gravitational/teleport/pull/17410#discussion_r997691681
+	entryMaxItems = 2000
+)
+
+type sharedDirectoryAuditCacheEntry struct {
+	nameCache         map[directoryID]directoryName
+	readRequestCache  map[completionID]readRequestInfo
+	writeRequestCache map[completionID]writeRequestInfo
+	// totalItems tracks the total number of items held in
+	// the entry between the nameCache, readRequestCache, and
+	// writeRequestCache.
+	totalItems uint32
+}
+
+func newSharedDirectoryAuditCacheEntry() *sharedDirectoryAuditCacheEntry {
+	return &sharedDirectoryAuditCacheEntry{
+		nameCache:         make(map[directoryID]directoryName),
+		readRequestCache:  make(map[completionID]readRequestInfo),
+		writeRequestCache: make(map[completionID]writeRequestInfo),
+		totalItems:        0,
+	}
+}
+
+// sharedDirectoryAuditCache is a data structure for caching information
+// from shared directory TDP messages so that it can be used later for
+// creating shared directory audit events.
+type sharedDirectoryAuditCache struct {
+	m map[sessionID]*sharedDirectoryAuditCacheEntry
+	sync.Mutex
+}
+
+func newSharedDirectoryAuditCache() sharedDirectoryAuditCache {
+	return sharedDirectoryAuditCache{
+		m: make(map[sessionID]*sharedDirectoryAuditCacheEntry),
+	}
+}
+
+// getInitialized gets an initialized sharedDirectoryAuditCacheEntry, mapped to sid.
+// If an entry at sid already exists, it returns that, otherwise it returns an empty, initialized entry.
+//
+// This should be called at the start of any SetX method to ensure that we never get a
+// "panic: assignment to entry in nil map".
+//
+// It is the responsibility of the caller to ensure that it has obtained the Lock before calling
+// getInitialized, and that it calls Unlock once the entry returned by getInitialized is no longer going to
+// be modified or otherwise used.
+func (c *sharedDirectoryAuditCache) getInitialized(sid sessionID) (entry *sharedDirectoryAuditCacheEntry) {
+	entry, ok := c.m[sid]
+
+	if !ok {
+		entry = newSharedDirectoryAuditCacheEntry()
+		c.m[sid] = entry
+	}
+
+	return entry
+}
+
+// SetName returns a non-nil error if the audit cache entry for sid exceeds its maximum size.
+// It is the responsiblity of the caller to terminate the session if a non-nil error is returned.
+func (c *sharedDirectoryAuditCache) SetName(sid sessionID, did directoryID, name directoryName) error {
+	c.Lock()
+	defer c.Unlock()
+
+	entry := c.getInitialized(sid)
+	if entry.totalItems >= entryMaxItems {
+		return trace.LimitExceeded("audit cache for sessionID(%v) exceeded maximum size", sid)
+	}
+
+	entry.nameCache[did] = name
+	entry.totalItems++
```

Comment:
Instead of having to keep the `totalItems` variable in sync with the actual total, have you considered a simple helper method?
```
func (s *sharedDirectoryAuditCacheEntry) totalItems() int {
   return len(s.nameCache) + len(s.readRequestCache) + len(s.writeRequestCache)
}
```

I like this better, as there's less of a chance for bugs related to forgetting to increment or decrement the counter.